src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,130 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private int size = 0;
+    private int capacity = DEFAULT_CAPACITY;
+    private Object[] array;
+
+    public ArrayList() {
+        array = new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        growIfArrayFull();
+        array[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        growIfArrayFull();
+        size++;
+        checkBounds(index);
 
+        for (int i = array.length - 1; i > index; i--) {
+            array[i] = array[i - 1];
+        }
+
+        array[index] = value;
     }
 
     @Override
     public void addAll(List<T> list) {
+        int listSize = list.size();
+        ensureCapacity(listSize);
+        int counter = 0;
+
+        for (int i = size; i < size + listSize; i++) {
+            array[i] = list.get(counter++);
+        }
+
+        size += listSize;
+    }
+
+    private void ensureCapacity(int listLength) {
+        while (!(capacity > size + listLength)) {
+            grow();
+        }
+    }
+
+    private void growIfArrayFull() {
+        if (size == capacity) {
+            grow();
+        }
+    }
 
+    private void grow() {
+        capacity = (int) (capacity * 1.5);
+        array = Arrays.copyOf(array, capacity);
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkBounds(index);
+        return (T) array[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkBounds(index);
+        array[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkBounds(index);
+        size--;
+        T objectRemoved = (T) array[index];
+
+        for (int i = index; i < array.length - 1; i++) {
+            array[i] = array[i + 1];
+        }
+
+        array[array.length - 1] = null;
+        return objectRemoved;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        T itemRemoved = null;
```

Comment:
You can simplify this method by
1. checking for index of element (if already present)
2. reusing `remove(int index)`